File: code/0144 Binary Tree Preorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Return an array of size *returnSize.
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* preorderTraversal(struct TreeNode* root, int* returnSize) {
   
    int *result;
    int top, count;
    struct TreeNode **stack;    /* stack points to the node */
    struct TreeNode  *current;  /* current points to the node's value */
    
    result = malloc(100 * sizeof(int));
    stack  = malloc(100 * sizeof(struct TreeNode *));
    top    = 0;
    count  = 0;
    current = root;
    
    while (top > 0 || current != NULL) {  
        while (current != NULL) {
            result[count++] = current->val;
            stack[top++]    = current;
            current = current->left;
        }
        
        current = stack[--top]; 
        /* Here we have arrived the leaf node, so the next step is to go back and search the right one */
        /* The feature of stack will help us to reach this goal easily */
        if (current->right == NULL) {
           current = NULL;
        } else {
           current = current->right;
        }
    }
    
    *returnSize = count;
    return (result);
}

File: Python/0144 Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    # recursively
    def recur(self, root, res):
        if root:
            res.append(root.val)
            self.recur(root.left,res)
            self.recur(root.right,res)
        
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res=[]
        self.recur(root,res)
        return res
        
    # iteratively
    def preorderTraversal(self, root):
        stack, res = [root], []
        while stack:
            node = stack.pop()
            if node:
                res.append(node.val)
                stack.append(node.right) # first in last out, so we first append 'right'
                stack.append(node.left)
        return res
        
